prodigy_demo/llm_recipe/llm_recipe.py

```diff
@@ -0,0 +1,137 @@
+"""
+A custom LLM recipe for text classification.
+
+prodigy textcat_llm llm_tweets_annotated \
+    ./data/data_to_label_500.jsonl \
+    -F llm_recipe.py
+"""
+from typing import Iterator
+import copy
+
+import prodigy
+from prodigy.components.loaders import JSONL
+from prodigy.components.preprocess import add_tokens
+import spacy
+import os
+
+from dotenv import load_dotenv
+
+import langchain_core.runnables.base
+
+import langchain_openai
+
+from langchain_core.prompts import ChatPromptTemplate
+from langchain_core.output_parsers import StrOutputParser
+
+
+load_dotenv()
+# here, we're loading our openai api key from a .env file
+# in the /llm_recipe directory
+oi_key = os.getenv('OPENAI_KEY')
+
+# if it doesn't exist, we'll throw an error
+if not oi_key:
+    assert False, "OPENAI_KEY not set in .env file"
+
+# let's instantiate our output parser so that the llm model
+# returns string rather than a ChatMessage object
+output_parser = StrOutputParser()
+
+# finally, lets instantiate our llm model - we're using OpenAI's
+llm = langchain_openai.ChatOpenAI(openai_api_key=oi_key)
+
+# The below function makes use of the llm,
+# output parser and prompt to make a chain of runnables
+# that we will invoke to return a llm-generated label
+
+
+def _make_chain(output_parser: langchain_core.output_parsers = output_parser,
+                llm: langchain_openai.ChatOpenAI = llm
+                ) -> langchain_core.runnables.base.RunnableSequence:
+    """Make a chain of runnables for text classification.
+
+    Args:
+        output_parser (langchain_core.output_parsers): Output parser
+        llm (langchain_openai.ChatOpenAI): LLM model
+
+    Returns:
+        langchain_core.runnables.base.RunnableSequence: Chain of runnables
+    """
+    prompt = ChatPromptTemplate.from_messages([
+        ("system", "You are world class data labeller. You will label the input sentence for sentiment as either positive or negative. Please only return the label itself."),
+        ("user", "{input_sentence}")
+    ])
+
+    chain = prompt | llm | output_parser
```

Comment:
what exactly is happening here?

Reply:
we're "chaining" together the prompt llm and output parser together such that you have one object that invokes all three - will add that as a comment